### DISCUSSION TITLE
Question and Suggestions

### DISCUSSION THREAD

**Discussion by instaloadererror** [nianeyna/ao3downloader#83] [Ideas]
Is there any way to get the downloader to create a .csv or something that has information on fics? The downloader is great for archival purposes, but once the stories are downloaded discoverability is near zero. You have to already know something exists to find it and some information is lost. I'd really love to have some kind of information page that I could use that would have data like: fic name, author, fandom(s), word count, kudos, hits, etc. Basically just the basic info from the main information panel given by Ao3. I tried the "l" option but it does not include kudos or hits. Kudos specifically is probably the most important sorting stat on the site so I'd really like to be able to get it.

Suggestions:

I don't know the etiquette for suggestions here because I'm not a coder, but I understand and appreciate that this is a free resource you've generously provided us and I'm not expecting anything of you. I just want to put these suggestions here on the off chance you're working on this in the future and one of these strikes you as useful. 

1) If the answer to my above question is "no" implementing that would be hugely useful.

2) When downloading it would be really useful if it were possible to specify a download folder. It takes a while to move tens of thousands of small files around. 

3) I haven't yet used the feature that searches for updates, but based on my understanding of how it works I have some suggestions. The first thing that comes to mind is that it seems like it would be beneficial if I could input a link and a folder location and have the downloader go through all fics from that link and check if they've been updated compared to what's currently in the specified folder. The biggest barrier for this downloader is that Ao3 rate limiter so anything that avoids opening unnecessary pages would speed things up a lot. This would also have the benefit of simultaneously downloading new stories. Another benefit of this is that it would remove the need of shuffling folders back and forth into the "downloads" folder that currently exists every time I want to download the new stories from a link. 

4) I could be wrong, but my understanding of the way the downloader checks for story updates seems imperfect. Based on what I saw posted it seems like the downloader checks if the chapters in a file is x/y where x=y and if so doesn't check for updates. It would be better for the purposes of archiving if there was an option to have it check every fic to see if x=a where the website had the chapters as a/b. 

5) Implement some kind of save feature associating a specific link with a specified folder for automatic updating across multiple folders at a time.

**Comment by nianeyna** [comment]
I am always fine with suggestions! If I don't want to do the suggestion I will simply not do it. I won't be offended if you aren't. 😉

For organizing fic once it is downloaded I highly recommend Calibre. I personally use it without plugins, but there are some available that are meant to optimize it for fanfic. I am aware of the [Easy Fanfic Library](https://github.com/KristalC/Easy-Fanfic-Library) project, though I have not used it.

You are the second person to ask about kudos. I feel some resistance to including metrics like kudos and hits in the downloader for two reasons:

1. I find sorting fics by popularity somewhat philosophically distasteful
2. The number of kudos and hits on a fic changes constantly, so recording the number of kudos on the specific day you happened to download a fic seems arbitrary to me, anyway

I'm not ruling this out entirely - I will continue to think about it. And of course, anyone is welcome to make their own fork that does whatever they want. But for now, it's not a priority for me.

I don't allow specifying your own download folder for security reasons, which are that I'm uneasy letting my program make changes to arbitrary folders on other people's computers. Like, it _probably_ would be fine, sure. But there is also a chance that I could make a huge mess of someone's machine that way. Which I really don't want to risk!

I am a little confused about your process that involves moving files back into the downloads folder - can you elaborate on that a bit more?

> have it check every fic to see if x=a where the website had the chapters as a/b.

I don't know what this means.

**Reply by instaloadererror** [reply]
> 1. I find sorting fics by popularity somewhat philosophically distasteful

Unfortunately, kudos or other popularity metrics are the only way of sorting fics in any useful way. I can't hop on the HP page and just randomly choose one of 450k stories to read. It's not perfect but it does generally provide a rough filter for quality. The issue with kudos changing presents a minor issue for file names, but it's not the big of a deal with a .csv. It shouldn't be too hard for users to combine separate files and delete the duplicate with a lower count (or just not see it at all after sorting).


> I am a little confused about your process that involves moving files back into the downloads folder - can you elaborate on that a bit more?

So, say I've downloaded the entire Wednesday (TV 2022) fanfiction list from Ao3 (which I have). I want to update my archive to include new stories. As of right now I need to move the ~10k files from my first download back into the "downloads" folder so the downloader doesn't spend several days redownloading all the stories I already have and then move them back. I would have to do this for each individual sub-folder of my archive. This isn't a quick process. Moving that many files takes a long time and manually doing so for dozens of different links would be very tedious.

> > have it check every fic to see if x=a where the website had the chapters as a/b.
> 
> I don't know what this means.

If I have a fic downloaded right now and in the file it says I have x/y chapters the downloader just checks if x=y. If it does it considers the fic complete and won't check for updates, otherwise it thinks it's incomplete and will check for updates. On the site a fic will be listed as having a/b chapters. x, y, a, and b can all be different. x can equal y, but if x!=a then the story has since been updated even if it was once complete and if y!=b then the author has since decided the story will be getting more chapters even if they haven't yet updated it and it shouldn't be considered complete. A further issue that might come up is that if x>a (or the word count has decreased) something has been removed from the story and the archiver should avoid replacing the current file and instead preferably create a second file that in some way indicates that it is an edited version missing previous content.

**Reply by nianeyna** [reply]
> Unfortunately, kudos or other popularity metrics are the only way of sorting fics in any useful way.

Skill issue. 😇 

To get new fics in a tag you can simply sort by date updated and filter to only fics that were updated after the last time you downloaded. (Normally I would say get this date from the log file, but we've established yours is too large. I will work on a solution to that, but in the meantime, you could just note it down in a text file or something.) You may get a few dupes this way, but not so many that it will meaningfully affect the runtime.

The reason the updater only checks for updates on fics where x != y is because if I don't filter out completed works _before_ starting to scrape ao3, the downloader would have to make a request to ao3 for _every single work in the folder._ Because otherwise how could it possibly _know_ whether the author has changed their mind about the fic being complete? There simply is no practical way to check for updates on fics that the author has already marked complete unless you want to just do a full redownload - which is an option, of course.

I do already check for if the chapter count has decreased since the last download, and skip redownloading in that case.

**Reply by kyrielie** [reply]
~~I kinda have the same issue of not being able to tell which fics I have already downloaded— but since I download random collections (avoiding the need for sorting by kudos/bookmarks) and then dump them into calibre I want to try to find a way to not redownload.~~

~~Right now, I came up with a janky solution by using calibre to fetch all the links in the files I download, using a shell script to compare my input file with the already downloaded links file. Which doesn't really work if I'm downloading from a link, but most collections are small enough that it doesn't matter that often.~~

I just noticed 'configure ignore list (list of links to never try to download)' omg I can just use that! How did I not notice that...

My suggestion for dealing with a large fanfiction library is calibre, idk how else to do it honestly. You can fetch kudos and other metrics from within the file using a plugin called "Extract People and Other Metadata", which I plan to do at some point maybe.

I haven't even tried to deal with updating works, if I really want to read an update I'll figure it out when I get around to reading that work. On another tangent, when you import works into calibre be careful to check that "duplicates" are actually duplicates, not works with the same name by different authors. (Or works by the same name by the same author, why would you do this to me.) I have accidentally lost a bunch of works this way.

**Comment by instaloadererror** [comment]
> the downloader would have to make a request to ao3 for every single work in the folder.

Could it work if you could include a link? I haven't actually used the update function yet because I'm still in the middle of my first pass archive, but if the downloader were to ask for a link to compare to the fics in the folder and crawl the link that would make it significantly faster right? Like it looks at the first story, if it's not in the folder it downloads it, if it is it checks the chapter info to see if it's been updated. 

> I do already check for if the chapter count has decreased since the last download, and skip redownloading in that case.

That's great. It's reassuring to know the downloader won't accidentally delete something in that case. I was kind of worried about that actually.

**Reply by nianeyna** [reply]
The point of the updater is _only_ to update fics that are already in your library, and only if they are incomplete. If you want to check for updates in a _tag_, you should do the date filter I described above.

**Reply by instaloadererror** [reply]
Here is another suggestion. I don't know exactly how the Ao3 rate limiter works so this might not be possible, but it would be cool if there was an option to keep the downloader a few below that rather than going until Ao3 tells it to stop, that way the downloader can be run constantly instead of having to be stopped during the time of day when I'm home and will be using the website.

**Reply by nianeyna** [reply]
Yeah, that was how ao3downloader quote-unquote "worked" for a while lol buuuuut in practice it just hit the rate limit all the time anyway 😔 so you basically had the worst of both worlds, you were getting rate-limited a lot, but you also were downloading much slower than you actually needed to. Reason being the ao3 rate limit is actually not consistent across the site, it is different per-page and I believe also may be different depending on which load-balancer node you happen to get on the backend. This logic is also all controlled by a config file which is not publicly available, and could change at any time. So, yeah, maybe if I were to work really hard on it, and harass an ao3 dev into giving me prod secrets, I could reverse-engineer something that would work maybe 50% of the time... XD

tl;dr yeah, not possible 😭
